src/Hamiltonians/ParitySymmetry.jl

```diff
@@ -0,0 +1,126 @@
+"""
+    left_weight(address)
+
+Compute how left-heavy the `address` is. A `left_weight` of 0 indicates that the address
+is symmetric. `left_weight(reverse(address)) = -left_weight(address)`.
+"""
+function left_weight(f::Union{BoseFS,FermiFS})
+    M = num_modes(f)
+    midpoint = cld(M, 2)
+    weight = 0
+    for index in occupied_modes(f)
+        weight += (index.mode - midpoint) * index.occnum
+    end
+    return weight
+end
+function left_weight(f::CompositeFS)
+    return _left_weight(f.components)
+end
+@inline _left_weight(::Tuple{}) = 0
+@inline function _left_weight((c, cs...))
+    weight = left_weight(c)
+    if weight == 0
+        return _left_weight(cs)
+    else
+        return weight
+    end
+end
+
+"""
+    ParitySymmetry(ham::AbstractHamiltonian{T}; even=true) <: AbstractHamiltonian{T}
+
+Changes the [`offdiagonals`](@ref) of a Hamiltonian in a way that produces even or odd
+eigenfunctions. For some Hamiltonians, this reduces the size of the Hilbert space by half.
+
+Whether the even or odd part is to be used is controlled by the keyword argument `even`.
+
+# Notes
+
+* This modifier only works on addresses with an even number of modes.
+* For the odd parity symmetry, the starting address of the underlying Hamiltonian can not be
+  symmetric.
+
+```jldoctest
+julia> ham = HubbardReal1D(BoseFS((0,2,1)))
+HubbardReal1D(BoseFS{3,3}((0, 2, 1)); u=1.0, t=1.0)
+
+julia> size(Matrix(ham))
+(10, 10)
+
+julia> size(Matrix(ParitySymmetry(ham)))
+(6, 6)
+
+julia> size(Matrix(ParitySymmetry(ham; odd=true)))
+(4, 4)
+
+julia> eigvals(Matrix(ham))[1]
+-5.095932280272496
+
+julia> eigvals(Matrix(ParitySymmetry(ham)))[1]
+-5.095932280272496
+```
+"""
+struct ParitySymmetry{T,H<:AbstractHamiltonian{T}} <: AbstractHamiltonian{T}
+    hamiltonian::H
+    even::Bool
+    direction::Int8
+end
+
+function ParitySymmetry(hamiltonian; odd=false, even=!odd)
+    address = starting_address(hamiltonian)
+
+    if !isodd(num_modes(address))
+        error("Starting address must have an odd number of modes")
```

Comment:
```suggestion
        throw(ArgumentError("Starting address must have an odd number of modes"))
```
See, e.g. https://discourse.julialang.org/t/error-vs-throw-argumenterror/18915

Reply:
Actually, what is the problem with even number of modes. The midpoint would be a non-integer, but is there a real problem? Parity should still exist and make sense.

Reply:
Ok, I'll use `ArgumentError`s from now on. 

The problem with the number of modes is in momentum space, where reversing an address with an even number of modes gives you an address with a different momentum. The correct way to do it there would be to leave the last mode alone, but that would break real space. We could get around this if we could distinguish momentum space addresses from real space ones.